src/features/wrap.rs

```diff
@@ -0,0 +1,782 @@
+                &text[byte_split_pos..]
+            };
+            stack.push((style, next_line));
+
+            curr_line.push((S::default(), &wrap_symbol));
+            result.push(curr_line);
+
+            curr_line = vec![(S::default(), LINEPREFIX)];
+            curr_sum = LINEPREFIX.len();
+        }
+    }
+
+    if !curr_line.is_empty() || result.is_empty() {
+        result.push(curr_line);
+    }
+
+    if !stack.is_empty() {
+        // `unwrap()` is ok, the previous `if` ensured `result` is not empty
+        result.last_mut().unwrap().extend(stack.into_iter().rev());
+    }
+
+    result
+}
+
+pub fn wrap_text_and_style<'a, S>(
+    folded_vec: &mut Vec<Vec<(S, &'a str)>>,
+    input_vec: Vec<(S, &'a str)>,
+    must_wrap: bool,
+    wrap_symbol: &'a str,
+    line_width: usize,
+    max_lines: usize,
+) -> (usize, usize)
+where
+    S: Copy + Default + std::fmt::Debug,
+{
+    let size_prev = folded_vec.len();
+
+    if must_wrap {
+        folded_vec.append(&mut wrap_line(
+            input_vec.into_iter(),
+            &wrap_symbol,
+            line_width,
+            max_lines,
+        ));
+    } else {
+        folded_vec.push(input_vec.to_vec());
+    }
+
+    (size_prev, folded_vec.len())
+}
+
+#[allow(clippy::comparison_chain, clippy::type_complexity)]
+pub fn wrap_plusminus_block<'c: 'a, 'a, S>(
+    config: &'c Config,
+    minus: Vec<Vec<(S, &'a str)>>,
+    plus: Vec<Vec<(S, &'a str)>>,
+    alignment: &[(Option<usize>, Option<usize>)],
+    line_width: &LineWidthPM,
+    wrapinfo: PlusMinus<S, Vec<bool>>,
+) -> (
+    Vec<(Option<usize>, Option<usize>)>,
+    PlusMinus<(), Vec<State>>,
+    PlusMinus<S, Vec<Vec<(S, &'a str)>>>,
+)
+where
+    S: Copy + Default + std::fmt::Debug,
+{
+    let mut new_alignment = vec![];
+
+    let mut new_minus_state = vec![];
+    let mut new_plus_state = vec![];
+
+    let mut new_minus = vec![];
+    let mut new_plus = vec![];
+
+    let mut minus_iter = minus.into_iter();
+    let mut plus_iter = plus.into_iter();
+
+    let mut minus_must_wrap_iter = wrapinfo.minus.into_iter();
+    let mut plus_must_wrap_iter = wrapinfo.plus.into_iter();
+
+    macro_rules! assert_alignment {
+        ($have:tt, $expected:tt, $msg2:tt) => {
+            assert_eq!(*$have, $expected, "bad alignment index {}", $msg2);
+            $expected += 1;
+        };
+    }
+    let mut m_expected = 0;
+    let mut p_expected = 0;
+
+    // Process blocks according to alignment and build a new alignment.
+    // Wrapped lines get assigned the state HunkMinusWrapped/HunkPlusWrapped.
+    for (minus, plus) in alignment {
+        let (minus_extended, plus_extended) = match (minus, plus) {
+            (Some(m), None) => {
+                assert_alignment!(m, m_expected, "l (-)");
+
+                let (minus_start, extended_to) = wrap_text_and_style(
+                    &mut new_minus,
+                    minus_iter.next().expect("bad alignment l (-)"),
```

Comment:
I think this is going to sound a bit odd, but would you mind using `unwrap_or_else(|| panic!(...))` instead of `expect()`? It's a local style rule that I've been following.

I'm prepared to be shouted down  :), but the reason is that I personally really dislike the idiom `.expect("something is wrong")`. It seems to me like the only bad design decision I've encountered so far in Rust -- that clearly reads as "I expect something is wrong", when apparently the Rust language designers thought it would be understood as "I expect this operation to succeed but, if not, then something is wrong".